src/client/dfuse/dfuse_core.c

```diff
@@ -671,204 +756,178 @@ dfuse_set_default_cont_cache_values(struct dfuse_cont *dfc)
+					if (!uuid_is_null(dfcc->dfcc_uuid))
+						continue;
+					dfc->dfs_ino = dfcc->dfcc_ino;
+					break;
+				}
+				if (dfc->dfs_ino != 0)
+					break;
+			}
+			D_SPIN_UNLOCK(&dfuse_info->di_lock);
+		}
+
+		dfc->dfc_attr_timeout       = 307;
+		dfc->dfc_dentry_dir_timeout = 307;
+		dfc->dfc_ndentry_timeout    = 307;
+
+		rc = ival_add_cont_buckets(dfc);
+		if (rc != 0)
+			goto err_free;
 
-	} else if (*_dfc == NULL) {
-		char str[37];
-		int  dfs_flags = O_RDWR;
+	} else {
+		daos_cont_info_t        c_info = {};
+		struct dfuse_cont_core *dfcc;
+		int                     dfs_flags = O_RDWR;
 
 		dfc->dfs_ops = &dfuse_dfs_ops;
-		uuid_copy(dfc->dfs_cont, *cont);
-		uuid_unparse(dfc->dfs_cont, str);
-		rc = daos_cont_open(dfp->dfp_poh, str, DAOS_COO_RW, &dfc->dfs_coh, NULL, NULL);
+
+		rc = daos_cont_open(dfp->dfp_poh, label, DAOS_COO_RW, &dfc->dfs_coh, &c_info, NULL);
 		if (rc == -DER_NO_PERM) {
 			dfs_flags = O_RDONLY;
-			rc = daos_cont_open(dfp->dfp_poh, str, DAOS_COO_RO, &dfc->dfs_coh, NULL,
-					    NULL);
+
+			rc = daos_cont_open(dfp->dfp_poh, label, DAOS_COO_RO, &dfc->dfs_coh,
+					    &c_info, NULL);
 		}
-		if (rc == -DER_NONEXIST) {
-			DFUSE_TRA_INFO(dfc, "daos_cont_open() failed: " DF_RC, DP_RC(rc));
-			D_GOTO(err_free, rc = daos_der2errno(rc));
-		} else if (rc != -DER_SUCCESS) {
-			DFUSE_TRA_ERROR(dfc, "daos_cont_open() failed: " DF_RC, DP_RC(rc));
+		if (rc != -DER_SUCCESS) {
+			if (rc == -DER_NONEXIST || rc == -DER_NO_PERM)
+				DHL_INFO(dfc, rc, "daos_cont_open() failed");
+			else
+				DHL_ERROR(dfc, rc, "daos_cont_open() failed");
 			D_GOTO(err_free, rc = daos_der2errno(rc));
 		}
+
 		rc = dfs_mount(dfp->dfp_poh, dfc->dfs_coh, dfs_flags, &dfc->dfs_ns);
 		if (rc) {
 			DHS_ERROR(dfc, rc, "dfs mount() failed");
 			D_GOTO(err_close, rc);
 		}
 
+		uuid_copy(dfc->dfc_uuid, c_info.ci_uuid);
+
 		if (dfuse_info->di_caching) {
 			rc = dfuse_cont_get_cache(dfc);
 			if (rc == ENODATA) {
-				/* If there are no container attributes then use defaults */
 				DFUSE_TRA_INFO(dfc, "Using default caching values");
 				dfuse_set_default_cont_cache_values(dfc);
-				rc = 0;
 			} else if (rc != 0) {
 				D_GOTO(err_umount, rc);
 			}
-
-		} else {
-			DFUSE_TRA_INFO(dfc, "Caching disabled");
 		}
 
 		rc = ival_add_cont_buckets(dfc);
 		if (rc != 0)
 			goto err_umount;
 
-	} else {
-		/* This is either a container where a label is set on the
-		 * command line, or one created through mkdir, in either case
-		 * the container will be mounted, and caching etc already
-		 * setup.
+		/* Check if this container has been accessed in the past and if so then reuse the
+		 * inode number.
 		 */
-		dfc->dfs_ops = &dfuse_dfs_ops;
+		D_SPIN_LOCK(&dfuse_info->di_lock);
+		d_list_for_each_entry(dfcc, &dfp->dfp_historic, dfcc_entry) {
+			if (dfcc->dfcc_ino == 0)
+				continue;
+			if (uuid_compare(dfcc->dfcc_uuid, dfc->dfc_uuid) != 0)
+				continue;
+			dfc->dfs_ino = dfcc->dfcc_ino;
+			break;
+		}
+		if (dfc->dfs_ino == 0) {
```

Comment:
not quite sure what this block is doing. it was not found in the historic list, what is the next list (di_pool_historic) for?

Reply:
It searches the list of historic container for the current pool and then if not found there then it needs to look at the historic pools.

If you connected to p1, opened and closed c1 then c1 would be on the open pool historic list.

If you connected to p1, opened and closed c1, disconnected from p1, connected to p1 and opened c1 then c1 would not be in the historic list for the open pool descriptor but would be on the historic list.  On p1 being closed if it's already on the historic pool list then it splices the list from the p1 descriptor it's closing to the p1 descriptor on the historic list.

This is actually the exact case the new (but disabled) test case covers.

Reply:
got it